From ed1c17fe3c1b2e7bff54727f612c7c56a3d37445 Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Date: Thu, 27 Dec 2018 16:27:28 +0200
Subject: [PATCH 007/269] s32g: Set SMPEN bit on U-Boot's behalf

U-Boot at EL2 will crash attempting to execute an EL3 instruction.

Upstream-Status: Pending 

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
Issue: ALB-3452
---
 plat/s32g/include/platform_def.h |  1 +
 plat/s32g/s32g275_bl31.c         |  2 +-
 plat/s32g/s32g_helpers.S         | 12 ++++++++++++
 3 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/plat/s32g/include/platform_def.h b/plat/s32g/include/platform_def.h
index 3d29607ed..3253c309a 100644
--- a/plat/s32g/include/platform_def.h
+++ b/plat/s32g/include/platform_def.h
@@ -153,5 +153,6 @@ int console_s32g_register(uintptr_t baseaddr, uint32_t clock, uint32_t baud,
 int console_s32g_putc(int c, struct console_s32g *console);
 int console_s32g_flush(struct console_s32g *console);
 int xrdc_enable(void *xrdc_addr);
+void s32g_smp_fixup(void);
 #endif /* __ASSEMBLY__ */
 #endif /* PLATFORM_DEF_H */
diff --git a/plat/s32g/s32g275_bl31.c b/plat/s32g/s32g275_bl31.c
index b7360847c..d2143e09a 100644
--- a/plat/s32g/s32g275_bl31.c
+++ b/plat/s32g/s32g275_bl31.c
@@ -81,7 +81,7 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
 
 void bl31_plat_arch_setup(void)
 {
-	/* TODO implement this */
+	s32g_smp_fixup();
 }
 
 void bl31_platform_setup(void)
diff --git a/plat/s32g/s32g_helpers.S b/plat/s32g/s32g_helpers.S
index caa044822..2daaa801b 100644
--- a/plat/s32g/s32g_helpers.S
+++ b/plat/s32g/s32g_helpers.S
@@ -22,6 +22,18 @@
  * Almost all these helper functions have been copied copied from
  * imx8_helpers.S (plat/imx), ls_helpers.S (plat/layerscape), or u-boot.
  */
+.globl s32g_smp_fixup
+
+
+/* Set SMPEN bit on u-boot's behalf */
+func s32g_smp_fixup
+	mrs x14, S3_1_c15_c2_1
+	orr x14, x14, #(1 << 6)
+	msr S3_1_c15_c2_1, x14
+	isb
+
+	ret
+endfunc s32g_smp_fixup
 
 /* Clobber list: x0,x1,x7,x8
  */
-- 
2.17.1

